lib/reservation_tracker.rb

```diff
@@ -0,0 +1,90 @@
+require 'date'
+require 'pry'
+
+class ReservationTracker
+  attr_reader :rooms, :all_reservations, :cost
+
+
+  def initialize
+    @all_reservations = []
+    @rooms = Room.new
+    @all_blocks = []
+    @block_room_reservations = []
+  end
+
+
+  def new_reservation(date_range)
+    date_range = Dates::date_range_format(date_range)
+    room_num = @rooms.find_available_room(occupied_rooms(date_range))
+    rate = @rooms.rate
+    reservation = Reservation.new(date_range, room_num, rate)
+    @all_reservations << reservation
+    return reservation
+  end
+
+
+  def reservation_list_by_date(date)
+    date = Dates::date_format(date)
+    reservation_list_by_date = @all_reservations.find_all { |reservation|
+      date >= reservation.date_range[:begin_date] && date < reservation.date_range[:end_date] }
+    return reservation_list_by_date
+  end
+
+
+  def occupied_rooms(date_range)
+    occupied_rooms = []
+    date_range = Dates::date_range_format(date_range)
+    @all_reservations.each do |reservation|
+      if Dates::date_range_comparison(date_range, reservation.date_range)
+        if reservation.room_num.class == Array
+          reservation.room_num.each do |num|
+            occupied_rooms << num
+          end
```

Comment:
This is a minor example of tight coupling: the `ReservationTracker` class needs to know a lot about the internals of `Reservation` in order to extract the room numbers. A better option might be to write a `get_room_numbers` method on `Reservation`, which always returns the numbers as an array.